File: Emulator/src/main/content/global/skill/summoning/familiar/Forager.java
package content.global.skill.summoning.familiar;

import core.game.node.entity.combat.equipment.WeaponInterface;
import core.game.node.entity.player.Player;
import core.game.node.item.Item;
import core.game.world.GameWorld;
import core.tools.RandomFunction;

/**
 * The type Forager.
 */
public abstract class Forager extends BurdenBeast {

    private final Item[] items;

    private int passiveDelay;

    /**
     * Instantiates a new Forager.
     *
     * @param owner       the owner
     * @param id          the id
     * @param ticks       the ticks
     * @param pouchId     the pouch id
     * @param specialCost the special cost
     * @param attackStyle the attack style
     * @param items       the items
     */
    public Forager(Player owner, int id, int ticks, int pouchId, int specialCost, int attackStyle, final Item... items) {
        super(owner, id, ticks, pouchId, specialCost, 30, attackStyle);
        this.items = items;
        setRandomPassive();
    }

    /**
     * Instantiates a new Forager.
     *
     * @param owner       the owner
     * @param id          the id
     * @param ticks       the ticks
     * @param pouchId     the pouch id
     * @param specialCost the special cost
     * @param items       the items
     */
    public Forager(Player owner, int id, int ticks, int pouchId, int specialCost, final Item... items) {
        this(owner, id, ticks, pouchId, specialCost, WeaponInterface.STYLE_DEFENSIVE, items);
    }

    @Override
    public void handleFamiliarTick() {
        super.handleFamiliarTick();
        if (items != null && items.length > 0 && passiveDelay < GameWorld.getTicks()) {
            if (RandomFunction.random(getRandom()) < 4) {
                produceItem(items[RandomFunction.random(items.length)]);
            }
            setRandomPassive();
        }
    }

    /**
     * Produce item boolean.
     *
     * @param item the item
     * @return the boolean
     */
    public boolean produceItem(final Item item) {
        if (!container.hasSpaceFor(item)) {
            owner.getPacketDispatch().sendMessage("Your familar is too full to collect items.");
            return false;
        }
        owner.getPacketDispatch().sendMessage(item.getAmount() == 1 ? "Your familar has produced an item." : "Your familiar has produced items.");
        return container.add(item);
    }

    /**
     * Produce item boolean.
     *
     * @return the boolean
     */
    public boolean produceItem() {
        if (items == null || items.length == 0) {
            return false;
        }
        return produceItem(items[RandomFunction.random(items.length)]);
    }

    /**
     * Handle passive action.
     */
    public void handlePassiveAction() {

    }

    /**
     * Gets random.
     *
     * @return the random
     */
    public int getRandom() {
        return 11;
    }

    /**
     * Sets random passive.
     */
    public void setRandomPassive() {
        passiveDelay = GameWorld.getTicks() + RandomFunction.random(100, 440);
    }
}

File: Emulator/src/main/content/global/skill/summoning/familiar/Forager.kt
package content.global.skill.summoning.familiar

import core.game.node.entity.combat.equipment.WeaponInterface
import core.game.node.entity.player.Player
import core.game.node.item.Item
import core.game.world.GameWorld
import core.tools.RandomFunction

/**
 * A familiar that forages and adds items to its inventory.
 *
 * @property items the list of possible items this familiar can forage
 */
abstract class Forager : BurdenBeast {

    private val items: Array<Item>
    private var passiveDelay: Int = 0

    /**
     * Constructs a new [Forager] with a custom attack style.
     *
     * @param owner the player who owns this familiar
     * @param id the NPC id of the familiar
     * @param ticks the active duration of the familiar in game ticks
     * @param pouchId the id of the summoning pouch used
     * @param specialCost the cost of using special abilities
     * @param attackStyle the combat style of the familiar
     * @param items the list of items it may forage
     */
    constructor(
        owner: Player?,
        id: Int,
        ticks: Int,
        pouchId: Int,
        specialCost: Int,
        attackStyle: Int,
        vararg items: Item
    ) : super(owner, id, ticks, pouchId, specialCost, 30, attackStyle) {
        this.items = Array(items.size) { i -> items[i] }
        setRandomPassive()
    }

    /**
     * Constructs a new [Forager] with the default DEFENSIVE attack style.
     */
    constructor(
        owner: Player,
        id: Int,
        ticks: Int,
        pouchId: Int,
        specialCost: Int,
        vararg items: Item
    ) : this(owner, id, ticks, pouchId, specialCost, WeaponInterface.STYLE_DEFENSIVE, *items)

    override fun handleFamiliarTick() {
        super.handleFamiliarTick()
        if (items.isNotEmpty() && passiveDelay < GameWorld.ticks) {
            if (RandomFunction.random(getRandom()) < 4) {
                produceItem(items[RandomFunction.random(items.size)])
            }
            setRandomPassive()
        }
    }

    /**
     * Attempts to produce and store the given item in the familiar's container.
     *
     * @param item the item to produce
     * @return `true` if added successfully, `false` otherwise
     */
    fun produceItem(item: Item): Boolean {
        if (!container.hasSpaceFor(item)) {
            owner.packetDispatch.sendMessage("Your familiar is too full to collect items.")
            return false
        }
        owner.packetDispatch.sendMessage(
            if (item.amount == 1)
                "Your familiar has produced an item."
            else
                "Your familiar has produced items."
        )
        return container.add(item)
    }

    /**
     * Produces a random item from the familiar's item list.
     *
     * @return `true` if added successfully, `false` otherwise
     */
    fun produceItem(): Boolean {
        if (items.isEmpty()) return false
        return produceItem(items[RandomFunction.random(items.size)])
    }

    /**
     * Hook for subclasses to implement additional passive behavior.
     */
    open fun handlePassiveAction() {
        // Optional override for custom behavior.
    }

    /**
     * Returns the random threshold used for passive item production checks.
     */
    open fun getRandom(): Int = 11

    /**
     * Sets the next passive item production delay.
     */
    private fun setRandomPassive() {
        passiveDelay = GameWorld.ticks + RandomFunction.random(100, 440)
    }
}
